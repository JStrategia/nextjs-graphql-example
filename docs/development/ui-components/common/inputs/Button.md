# Tooltip

Wraps children (normally only text) in a stylized button. Uses a [tooltip](../data/Tooltip.md) for accessibility.

## Props

- `children`: Wrapped React node
- `title`: Tooltip text
- `disabled`: Makes button unusable (default = true)
- `onClick`: Function triggered when button is clicked
- `type`: Button type ('button', 'submit' or 'reset') (default = 'button')
- `variant`: Button style variant ('contained', 'outlined' or 'text') (default = 'outlined')

## Usage

```tsx
function MyComponent() {
  return (
    <Button title="Some text that will be displayed on hover">
      <Body1>Some text</Body1>
    </Button>
  );
}
```
