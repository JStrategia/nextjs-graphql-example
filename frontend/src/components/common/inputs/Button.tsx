import React from 'react';
import styled, { css } from 'styled-components';
import Tooltip from 'components/common/data/Tooltip';

type ButtonType = 'button' | 'submit' | 'reset';

type Variant = 'contained' | 'outlined' | 'text';

type ButtonProps = {
  readonly children: React.ReactNode;
  readonly title: string;
  readonly disabled?: boolean;
  readonly onClick?: React.MouseEventHandler<HTMLAnchorElement | HTMLButtonElement>;
  readonly type?: ButtonType;
  readonly variant?: Variant;
};

function Button({ title, children, onClick, disabled = false, type = 'button', variant = 'outlined' }: ButtonProps) {
  return (
    <Tooltip title={title}>
      <div>
        <StyledButton type={type} variant={variant} onClick={onClick} disabled={disabled}>
          {children}
        </StyledButton>
      </div>
    </Tooltip>
  );
}

const ButtonTypography = css(
  ({ theme }) => css`
    font-family: sans-serif;
    font-weight: ${theme.fonts.weights.regular};
    font-size: ${theme.fonts.sizes.s};
    text-transform: uppercase;
    line-height: 1.75;
  `,
);

const StyledButton = styled.button<{ readonly variant: Variant }>(
  ({ variant, theme }) => css`
    ${ButtonTypography};
    display: flex;
    border-radius: ${theme.spacing.xxs};
    cursor: pointer;
    padding: ${theme.spacing.xxs} ${theme.spacing.xs};
    outline: 0;
    border: 0;
    color: ${theme.colors.palette.primary.main};
    background-color: transparent;
    transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
      color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

    ${variant === 'contained' &&
    `
    color: ${theme.colors.onPrimary};
    background-color: ${theme.colors.palette.primary.main};
    
    :hover {
      background-color: ${theme.colors.palette.primary.light};
    }
  `}

    ${variant === 'outlined' &&
    `
    :hover {
      background-color: ${theme.colors.surface};
    }
  `}
  `,
);

export default Button;
