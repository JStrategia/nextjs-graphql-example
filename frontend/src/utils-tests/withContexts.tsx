import React from 'react';
import ThemeProvider from 'contexts/ThemeProvider';

function withContexts(children: React.ReactNode) {
  return <ThemeProvider>{children}</ThemeProvider>;
}

export default withContexts;
