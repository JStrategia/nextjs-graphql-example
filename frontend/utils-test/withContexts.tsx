import React, { ReactNode } from 'react';
import ThemeProvider from '../contexts/ThemeProvider';

function withContexts(children: ReactNode) {
  return <ThemeProvider>{children}</ThemeProvider>;
}

export default withContexts;
