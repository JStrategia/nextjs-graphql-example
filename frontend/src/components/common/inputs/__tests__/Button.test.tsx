import React from 'react';
import { create } from 'react-test-renderer';
import Button, { ButtonType, ButtonVariant } from 'components/common/inputs/Button';
import withContexts from 'utils-tests/withContexts';

describe('Button', () => {
  describe('Given title and children', () => {
    const title = 'some title';
    const children = <span>Children</span>;

    describe.each([
      ['button', 'contained', true],
      ['button', 'contained', false],
      ['button', 'outlined', true],
      ['button', 'outlined', false],
      ['button', 'text', true],
      ['button', 'text', false],
      ['submit', 'contained', true],
      ['submit', 'contained', false],
      ['submit', 'outlined', true],
      ['submit', 'outlined', false],
      ['submit', 'text', true],
      ['submit', 'text', false],
      ['reset', 'contained', true],
      ['reset', 'contained', false],
      ['reset', 'outlined', true],
      ['reset', 'outlined', false],
      ['reset', 'text', true],
      ['reset', 'text', false],
    ])('Given %s type, %s variant and %s disabled', (buttonType, buttonVariant, disabled) => {
      describe('When rendering', () => {
        const render = create(
          withContexts(
            <Button
              title={title}
              type={buttonType as ButtonType}
              variant={buttonVariant as ButtonVariant}
              disabled={disabled}
            >
              {children}
            </Button>,
          ),
        ).toJSON();

        test('Then match snapshot', () => {
          expect(render).toMatchSnapshot();
        });
      });
    });
  });
});
