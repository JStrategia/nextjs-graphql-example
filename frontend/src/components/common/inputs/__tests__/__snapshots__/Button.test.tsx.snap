// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Given title and children When rendering Then match snapshot 1`] = `
<div
  className="sc-gsnTZi eZrxQK"
>
  <div>
    <button
      className="sc-dkzDqf hKofOt"
      disabled={false}
      type="button"
    >
      <span>
        Children
      </span>
    </button>
  </div>
  <span
    className="sc-bczRLJ gnxjFk"
  >
    some title
  </span>
</div>
`;
